daemon/c_cgroups_sockopt.c

```diff
@@ -0,0 +1,327 @@
+
+	for (list_t *l = sockopt->bpf_progs; l; l = l->next) {
+		c_cgroups_sockopt_prog_t *prog = l->data;
+
+		union bpf_attr detach_attr = {
+			.attach_type = BPF_CGROUP_SETSOCKOPT,
+			.target_fd = cgroup_fd,
+			.attach_bpf_fd = prog->fd,
+		};
+
+		// if cgroup still exist detach the program from it
+		if (cgroup_fd > 0 && bpf(BPF_PROG_DETACH, &detach_attr, sizeof(detach_attr)))
+			WARN_ERRNO("Failed to detach bpf sock program!");
+
+		close(prog->fd);
+		c_cgroups_sockopt_prog_free(prog);
+	}
+
+	list_delete(sockopt->bpf_progs);
+	sockopt->bpf_progs = NULL;
+
+	close(cgroup_fd);
+}
+
+#define BPF_LOG_SIZE 1024 * 1024
+#define BPF_PROG_LOAD_RETRIES 10
+
+static int
+c_cgroups_sockopt_prog_activate(c_cgroups_sockopt_t *sockopt, c_cgroups_sockopt_prog_t *prog)
+{
+	ASSERT(sockopt);
+	ASSERT(prog);
+
+	int cgroup_fd = -1;
+
+	union bpf_attr load_attr = {
+		.prog_type = BPF_PROG_TYPE_CGROUP_SOCKOPT,
+		.insns = ptr_to_u64(prog->insn),
+		.insn_cnt = prog->insn_n_structs,
+		.license = ptr_to_u64("GPL"),
+		.expected_attach_type = BPF_CGROUP_SETSOCKOPT,
+	};
+
+	INFO("bpf insns: %d, %llx", load_attr.insn_cnt, load_attr.insns);
+
+	prog->fd = bpf(BPF_PROG_LOAD, &load_attr, sizeof(load_attr));
+
+	int retry = 0;
+	while (prog->fd < 0 && errno == EAGAIN && retry < BPF_PROG_LOAD_RETRIES) {
+		retry++;
+		TRACE_ERRNO("Failed to load bpf program retrying (retry %d)!", retry);
+		prog->fd = bpf(BPF_PROG_LOAD, &load_attr, sizeof(load_attr));
+	}
+
+	if (prog->fd < 0) {
+		WARN_ERRNO("Failed to load bpf program retrying with logbuffer!");
+		char *bpf_log = mem_new0(char, BPF_LOG_SIZE);
+		load_attr.log_buf = ptr_to_u64(bpf_log);
+		load_attr.log_size = BPF_LOG_SIZE;
+		load_attr.log_level = 1;
+		// try again to get log
+		prog->fd = bpf(BPF_PROG_LOAD, &load_attr, sizeof(load_attr));
+		if (prog->fd < 0) {
+			ERROR_ERRNO("Failed to load bpf program '%s'!", bpf_log);
+			mem_free0(bpf_log);
+			goto error;
+		}
+		mem_free0(bpf_log);
+	}
+
+	cgroup_fd = open(sockopt->path, O_DIRECTORY | O_RDONLY | O_CLOEXEC);
+
+	IF_TRUE_RETVAL(cgroup_fd < 0, -COMPARTMENT_ERROR_CGROUPS);
+
+	union bpf_attr attach_attr = {
+		.attach_type = BPF_CGROUP_SETSOCKOPT,
+		.target_fd = cgroup_fd,
+		.attach_bpf_fd = prog->fd,
+		.attach_flags = BPF_F_ALLOW_MULTI,
+	};
+
+	int ret = bpf(BPF_PROG_ATTACH, &attach_attr, sizeof(attach_attr));
+	if (ret) {
+		ERROR_ERRNO("Failed to attach bpf program!");
+		goto error;
+	}
+
+	sockopt->bpf_progs = list_append(sockopt->bpf_progs, prog);
+	close(cgroup_fd);
+
+	return 0;
+error:
+
+	close(prog->fd);
+	close(cgroup_fd);
+	c_cgroups_sockopt_prog_free(prog);
+	return -1;
+}
+
+static void *
+c_cgroups_sockopt_new(compartment_t *compartment)
+{
+	ASSERT(compartment);
+	IF_NULL_RETVAL(compartment_get_extension_data(compartment), NULL);
+
+	c_cgroups_sockopt_t *sockopt = mem_new0(c_cgroups_sockopt_t, 1);
+	sockopt->container = compartment_get_extension_data(compartment);
+
+	sockopt->path = mem_printf("%s/%s", c_cgroups_subtree,
```

Comment:
check if c_cgroups_subtree != NULL?
(e.g. to emit error on wrong ordering of modules in Makefile?)